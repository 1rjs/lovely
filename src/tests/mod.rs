use std::fs;

use insta::glob;

use crate::{checker::Checker, ir::IRGenerator, lexer::Lexer, parser::Parser};

#[test]
fn compiler_tests() {
    glob!("source_files/*.lv", |path| {
        let input = fs::read_to_string(path).unwrap();

        // test the lexer
        let lexer = Lexer::new(&input);
        let tokens = lexer.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);

        // test the parser
        let mut parser = Parser::new(&input);
        let ast = parser.parse().unwrap();
        insta::assert_debug_snapshot!(ast);

        // test the checker
        let mut checker = Checker::new();
        let checked_program = checker.check(&ast).unwrap();
        insta::assert_debug_snapshot!(checked_program);

        // test the IR
        let ir_generator = IRGenerator::new(checker.types);
        let ir = ir_generator.program_ir(&checked_program);
        let ir_string = ir
            .iter()
            .map(|b| b.to_string())
            .collect::<Vec<_>>()
            .join("\n");
        insta::assert_snapshot!(ir_string);
    });
}
