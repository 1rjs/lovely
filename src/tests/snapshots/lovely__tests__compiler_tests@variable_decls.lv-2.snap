---
source: src/tests/mod.rs
expression: ast
input_file: src/tests/source_files/variable_decls.lv
---
Program(
    [
        Expression {
            span: Span {
                start: 0,
                end: 84,
            },
            kind: VariableDecl {
                name: "main",
                value: Expression {
                    span: Span {
                        start: 8,
                        end: 84,
                    },
                    kind: Function {
                        parameters: [],
                        return_type: None,
                        body: Expression {
                            span: Span {
                                start: 16,
                                end: 84,
                            },
                            kind: Block(
                                [
                                    Expression {
                                        span: Span {
                                            start: 20,
                                            end: 28,
                                        },
                                        kind: VariableDecl {
                                            name: "foo",
                                            value: Expression {
                                                span: Span {
                                                    start: 27,
                                                    end: 28,
                                                },
                                                kind: IntLiteral(
                                                    6,
                                                ),
                                            },
                                            mutable: false,
                                            ty: None,
                                        },
                                    },
                                    Expression {
                                        span: Span {
                                            start: 32,
                                            end: 40,
                                        },
                                        kind: VariableDecl {
                                            name: "bar",
                                            value: Expression {
                                                span: Span {
                                                    start: 39,
                                                    end: 40,
                                                },
                                                kind: IntLiteral(
                                                    7,
                                                ),
                                            },
                                            mutable: true,
                                            ty: None,
                                        },
                                    },
                                    Expression {
                                        span: Span {
                                            start: 44,
                                            end: 61,
                                        },
                                        kind: VariableDecl {
                                            name: "baz",
                                            value: Expression {
                                                span: Span {
                                                    start: 57,
                                                    end: 61,
                                                },
                                                kind: BoolLiteral(
                                                    true,
                                                ),
                                            },
                                            mutable: true,
                                            ty: Some(
                                                Ident(
                                                    "Bool",
                                                ),
                                            ),
                                        },
                                    },
                                    Expression {
                                        span: Span {
                                            start: 65,
                                            end: 82,
                                        },
                                        kind: VariableDecl {
                                            name: "qux",
                                            value: Expression {
                                                span: Span {
                                                    start: 78,
                                                    end: 82,
                                                },
                                                kind: Unit,
                                            },
                                            mutable: false,
                                            ty: Some(
                                                Ident(
                                                    "Unit",
                                                ),
                                            ),
                                        },
                                    },
                                ],
                            ),
                        },
                    },
                },
                mutable: false,
                ty: None,
            },
        },
    ],
)
